src/sage/rings/lazy_species.py

```diff
@@ -0,0 +1,1395 @@
+                                         LazyPowerSeriesRing,
+                                         LazySymmetricFunctions)
+from sage.rings.species import PolynomialSpecies, _label_sets
+from sage.data_structures.stream import (Stream_zero,
+                                         Stream_exact,
+                                         Stream_truncated,
+                                         Stream_function)
+from sage.categories.tensor import tensor
+from sage.combinat.integer_vector import IntegerVectors
+from sage.combinat.subset import subsets
+from sage.combinat.sf.sf import SymmetricFunctions
+from sage.combinat.partition import Partitions, Partition
+from sage.combinat.permutation import CyclicPermutations
+from sage.combinat.set_partition import SetPartitions
+from sage.graphs.graph_generators import graphs
+from sage.groups.perm_gps.permgroup_named import SymmetricGroup, CyclicPermutationGroup
+from sage.structure.element import parent
+import itertools
+from collections import defaultdict
+
+
+def weighted_compositions(n, d, weight_multiplicities, _w0=0):
+    r"""
+    Return all compositions of `n` of weight `d`.
+
+    The weight of a composition `n_1, n_2, \dots` is `\sum_i w_i n_i`.
+
+    INPUT:
+
+    - ``n`` -- a nonnegative integer, the sum of the parts
+    - ``d`` -- a nonnegative integer, the total weight
+    - ``weight_multiplicities`` -- an iterable,
+      ``weight_multiplicities[i]`` is the number of positions with
+      weight `i+1`.
+
+    .. TODO::
+
+        Possibly this could be merged with
+        :class:`~sage.combinat.integer_vector_weighted.WeightedIntegerVectors`.
+        However, that class does not support fixing the sum of the
+        parts currently.
+
+    EXAMPLES::
+
+        sage: from sage.rings.lazy_species import weighted_compositions
+        sage: list(weighted_compositions(1, 1, [2,1]))
+        [[1, 0], [0, 1]]
+
+        sage: list(weighted_compositions(2, 1, [2,1]))
+        []
+
+        sage: list(weighted_compositions(1, 2, [2,1,1]))
+        [[0, 0, 1]]
+
+        sage: list(weighted_compositions(3, 4, [2,2]))
+        [[2, 0, 1, 0],
+         [1, 1, 1, 0],
+         [0, 2, 1, 0],
+         [2, 0, 0, 1],
+         [1, 1, 0, 1],
+         [0, 2, 0, 1]]
+
+    """
+    # the empty composition exists if and only if n == d == 0
+    if not n:
+        if not d:
+            yield []
+        return
+    if not d:
+        return
+
+    # otherwise we iterate over the possibilities for the first
+    # weight_multiplicities[_w0] parts
+    try:
+        if _w0 >= len(weight_multiplicities):
+            return
+    except TypeError:
+        pass
+    if _w0 > d:
+        return
+    for s in range(n + 1):
+        for c in weighted_compositions(n - s, d - s * (_w0 + 1), weight_multiplicities, _w0=_w0+1):
+            m = weight_multiplicities[_w0]
+            for v in map(list, IntegerVectors(s, length=m)):
+                yield v + c
```

Comment:
```suggestion
    from sage.combinat.integer_lists.invlex import IntegerListsBackend_invlex
    for s in range(n + 1):
        for c in weighted_compositions(n - s, d - s * (_w0 + 1), weight_multiplicities, _w0=_w0+1):
            m = weight_multiplicities[_w0]
            for v in IntegerListsBackend_invlex(s, length=m)._iter():
                yield v + c
```
This should be faster without heavy parent/element objects.